### PR TITLE
[DISCO-2517] Test Documentation

### PR DESCRIPTION
**Pull request by taddes**
## References

JIRA: [DISCO-2517](https://mozilla-hub.atlassian.net/browse/DISCO-2517)
GitHub: [#167 ](https://github.com/mozilla-services/consvc-shepherd/issues/167)

## Description
Add some test documentation. Not all of this applies to Shepherd (because it’s low priority). 

Some stuff to consider:

    High level strategy for service. Types of tests that are run in CI.

    Instructions on how to run tests locally

    Architecture diagram for more involved tests

    Ensure you can test in situ (e.g. can you easily test a mobile devices using a staging server without using a custom build? How hard is it to use this API with a scripting system or CURL? How secure is it from griefing attacks? etc.)

## PR Review Checklist

_Put an `x` in the boxes that apply_

- [ ] This PR conforms to the [Contribution Guidelines](https://github.com/mozilla-services/consvc-shepherd/blob/main/CONTRIBUTING.md).
- [ ] The PR title starts with the JIRA issue reference, format `[DISCO-####]`, and has the same title (if applicable).
- [ ] [Documentation](https://github.com/mozilla-services/consvc-shepherd/tree/main/docs) has been updated (if applicable).
- [ ] Functional and performance test coverage has been expanded and maintained (if applicable).

[DISCO-2517]: https://mozilla-hub.atlassian.net/browse/DISCO-2517?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: docs/testing.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+## Unit Tests
+
+The unit layer is suitable for testing complex behavior at a small scale, 
+with fine-grained control over the inputs. 
+Due to their narrow scope, unit tests are fundamental to thorough test coverage.
+
+Our test coverage minimum is set to 95%. 
+A coverage report is printed to the shell showing the test results and possible exceptions. See [coverage.py](https://coverage.readthedocs.io/en/latest/) for more details.
+
+Unit tests are written and executed with `pytest` and are located in the `consvc_shepherd/tests/`
```
Comment (tiftran) [inline docs/testing.md:14]: test directories maybe?

File: docs/testing.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+directory. The name of individual test files should match those of the files they are testing in the form `test*.py`.
+Ex. `models.py` should be `test_models.py`. 
+
+To execute unit tests, use: `make test`. By default, these tests run in CI during each push 
+to an active PR and during the merge process.
+
+## Testing a Django App
```
Comment (tiftran) [inline docs/testing.md:21]: Not sure about this section, there's also mentions of some things that aren't actually done in Shepherd like  `setUpTestData`. Personally, when trying to hop onto tests, I look at existing tests, to ensure they're written uniformly? What do you think?

File: docs/testing.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
+# Test Strategy
+
+Shepherd's test code is in the `consvc_shepherd/tests` directory.
```
[the remark this answers is not in the document]
Comment (taddes) [reply docs/testing.md:3]: Good point, I'll add those refs!

File: docs/testing.md (hunk 2 above)
Comment (taddes) [reply docs/testing.md:21]: Yeah, I can possibly remove some of the things we don't use, but when reading up on the testing methods and classes, I thought adding context to the 'auto-magical' Django methods like `setUpTestData` and `setUp` might mislead people if they've not worked with Django. It's easy to at first think these are just user-defined and not part of the django.test framework.  Anything specific you'd want me to take out or change?

Comment (tiftran) [reply docs/testing.md:20]: lets remove setupTestData and the Test Models, Test Forms and Test View sections.

Comment (taddes) [reply docs/testing.md:20]: Will do.

Comment (mtbottle) [reply docs/testing.md:20]: I'm going to strongly encourage not adding this section. 

If someone new to this service is looking at how tests are written, then going to the source of the code if probably the best idea. The specifics about how Django implements tests is best left to the official Django documentation. If Django changes the way they do X in the future, we're going to have to also maintain this which is not ideal.

File: docs/testing.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+Shepherd's test code is in the `consvc_shepherd/tests`, `contile/tests`, and `openidc/tests` directories.
+
+## Unit Tests
+
+The unit layer is suitable for testing complex behavior at a small scale, 
+with fine-grained control over the inputs. 
+Due to their narrow scope, unit tests are fundamental to thorough test coverage.
+
+Our test coverage minimum is set to 95%.
```
Comment (mtbottle) [inline docs/testing.md:11]: This is a piece of volatile data that I would also defer users to the source. So let's remove references to it here.

**Comment by taddes** [comment]
Revisions made based on your comments @mtbottle , thx!

**Review by mtbottle — DISMISSED** [review]
I've added a few suggestions. Also, I think it would be good to have the "strategy" be more prominent. For instance, we only care about functional unit tests (vs. integration or end-to-end or performance testing), so that should be highlighted at the top. I've got a suggestion below on a concise way to state that.

File: docs/testing.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+Due to their narrow scope, unit tests are fundamental to thorough test coverage.
+
+Our test coverage minimum is set to 95%. 
+A coverage report is printed to the shell showing the test results and possible exceptions. See [coverage.py](https://coverage.readthedocs.io/en/latest/) for more details.
+
+Unit tests are written and executed with `pytest` and are located in the test directories listed above, in the form: `*/tests/`. The name of individual test files should match those of the files they are testing in the form `test*.py`.
+Ex. `models.py` should be `test_models.py`. 
+
+To execute unit tests, use: `make test`. By default, these tests run in CI during each push 
+to an active PR and during the merge process.
```
[the remark this answers is not in the document]
Comment (taddes) [reply docs/testing.md:18]: Good call @mtbottle ! If it's ok with you, I added a link to the Shepherd CI dashboard, just like it's listed in our Tech Docs Hub. You like that addition or want me to leave it out?

File: docs/testing.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
+# Test Strategy
+
+Shepherd's test code is in the `consvc_shepherd/tests`, `contile/tests`, and `openidc/tests` directories.
+
+## Unit Tests
+
+The unit layer is suitable for testing complex behavior at a small scale, 
+with fine-grained control over the inputs. 
+Due to their narrow scope, unit tests are fundamental to thorough test coverage.
```
[the remark this answers is not in the document]
Comment (taddes) [reply docs/testing.md:9]: lgtm!

**Comment by taddes** [comment]
Great suggestions @mtbottle , thanks for taking the time to look. Pushed the changes and just added the CI dashboard link, though am not attached to it and can remove if you think it's not needed. Otherwise everything is as requested.

**Pull request merged**